<a href="https://colab.research.google.com/github/DuckofDoom91/Portfolio/blob/main/wrangle_act.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project: Wrangling and Analyze Data

## Data Gathering
In the cell below, gather **all** three pieces of data for this project and load them in the notebook. **Note:** the methods required to gather each data are different.
1. Directly download the WeRateDogs Twitter archive data (twitter_archive_enhanced.csv)

In [ ]:
import pandas as pd
import numpy as np
import requests
import json
import matplotlib.pyplot as plt
%matplotlib inline
pd.set_option('display.max_colwidth', None)

In [ ]:
twitter_archive = pd.read_csv('twitter-archive-enhanced.csv')
twitter_archive.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                                                               source  \
0  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
3  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
4  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                 text  \
0                                                       This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU   
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2                   This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB   
3                                                             This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ   
4  This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek https://t.co/AtUZn91f7f   

   retweeted_status_id  retweeted_status_user_id retweeted_status_timestamp  \
0                  NaN                       NaN                        NaN   
1                  NaN                       NaN                        NaN   
2                  NaN                       NaN                        NaN   
3                  NaN                       NaN                        NaN   
4                  NaN                       NaN                        NaN   

                                                                                                                     expanded_urls  \
0                                                                  https://twitter.com/dog_rates/status/892420643555336193/photo/1   
1                                                                  https://twitter.com/dog_rates/status/892177421306343426/photo/1   
2                                                                  https://twitter.com/dog_rates/status/891815181378084864/photo/1   
3                                                                  https://twitter.com/dog_rates/status/891689557279858688/photo/1   
4  https://twitter.com/dog_rates/status/891327558926688256/photo/1,https://twitter.com/dog_rates/status/891327558926688256/photo/1   

   rating_numerator  rating_denominator      name doggo floofer pupper puppo  
0                13                  10   Phineas  None    None   None  None  
1                13                  10     Tilly  None    None   None  None  
2                12                  10    Archie  None    None   None  None  
3                13                  10     Darla  None    None   None  None  
4                12                  10  Franklin  None    None   None  None

2. Use the Requests library to download the tweet image prediction (image_predictions.tsv)

In [ ]:
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)
response

<Response [200]>

In [ ]:
with open('image-predictions.tsv', mode = 'wb') as file:
    file.write(response.content)
image_pred = pd.read_csv('image-predictions.tsv', sep = '\t')

image_pred.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

3. Use the Tweepy library to query additional data via the Twitter API (tweet_json.txt)

In [ ]:
twitter_list = []
with open('tweet-json.txt', 'r') as file:
    lines = file.readlines()
    for line in lines:
        parsed_json = json.loads(line)
        twitter_list.append({'tweet_id': parsed_json['id'],
                        'retweet_count': parsed_json['retweet_count'],
                        'favorite_count': parsed_json['favorite_count']})
tweet_json = pd.DataFrame(twitter_list, columns = ['tweet_id', 'retweet_count', 'favorite_count'])

tweet_json.head()

tweet_id  retweet_count  favorite_count
0  892420643555336193           8853           39467
1  892177421306343426           6514           33819
2  891815181378084864           4328           25461
3  891689557279858688           8964           42908
4  891327558926688256           9774           41048

## Assessing Data
In this section, detect and document at least **eight (8) quality issues and two (2) tidiness issue**. You must use **both** visual assessment
programmatic assessement to assess the data.

**Note:** pay attention to the following key points when you access the data.

* You only want original ratings (no retweets) that have images. Though there are 5000+ tweets in the dataset, not all are dog ratings and some are retweets.
* Assessing and cleaning the entire dataset completely would require a lot of time, and is not necessary to practice and demonstrate your skills in data wrangling. Therefore, the requirements of this project are only to assess and clean at least 8 quality issues and at least 2 tidiness issues in this dataset.
* The fact that the rating numerators are greater than the denominators does not need to be cleaned. This [unique rating system](http://knowyourmeme.com/memes/theyre-good-dogs-brent) is a big part of the popularity of WeRateDogs.
* You do not need to gather the tweets beyond August 1st, 2017. You can, but note that you won't be able to gather the image predictions for these tweets since you don't have access to the algorithm used.


## `twitter_archive` Assesment

In [ ]:
twitter_archive.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2330  666362758909284353                    NaN                  NaN   
1835  676101918813499392                    NaN                  NaN   
657   791774931465953280                    NaN                  NaN   
1788  677557565589463040                    NaN                  NaN   
2008  672256522047614977                    NaN                  NaN   

                      timestamp  \
2330  2015-11-16 21:10:36 +0000   
1835  2015-12-13 18:10:33 +0000   
657   2016-10-27 22:53:48 +0000   
1788  2015-12-17 18:34:46 +0000   
2008  2015-12-03 03:30:19 +0000   

                                                                                  source  \
2330  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1835  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
657                      <a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>   
1788  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2008  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                              text  \
2330  Unique dog here. Very small. Lives in container of Frosted Flakes (?). Short legs. Must be rare 6/10 would still pet https://t.co/XMD9CwjEnM   
1835                                                                         Meet Striker. He's ready for Christmas. 11/10 https://t.co/B3xxSLjQSH   
657                                                     Vine will be deeply missed. This was by far my favorite one. 14/10 https://t.co/roqIxCvEB3   
1788  Reckless pupper here. Not even looking at road. Absolute menace. No regard for fellow pupper lives. 10/10 still cute https://t.co/96IBkOYB7j   
2008   Mighty rare dogs here. Long smooth necks. Great knees. Travel in squads. 1 out of every 14 is massive. 8/10 for all https://t.co/PoMKKnKpRd   

      retweeted_status_id  retweeted_status_user_id  \
2330                  NaN                       NaN   
1835                  NaN                       NaN   
657                   NaN                       NaN   
1788                  NaN                       NaN   
2008                  NaN                       NaN   

     retweeted_status_timestamp  \
2330                        NaN   
1835                        NaN   
657                         NaN   
1788                        NaN   
2008                        NaN   

                                                        expanded_urls  \
2330  https://twitter.com/dog_rates/status/666362758909284353/photo/1   
1835  https://twitter.com/dog_rates/status/676101918813499392/photo/1   
657                                     https://vine.co/v/ea0OwvPTx9l   
1788  https://twitter.com/dog_rates/status/677557565589463040/photo/1   
2008  https://twitter.com/dog_rates/status/672256522047614977/photo/1   

      rating_numerator  rating_denominator     name doggo floofer  pupper  \
2330                 6                  10     None  None    None    None   
1835                11                  10  Striker  None    None    None   
657                 14                  10     None  None    None    None   
1788                10                  10     None  None    None  pupper   
2008                 8                  10     None  None    None    None   

     puppo  
2330  None  
1835  None  
657   None  
1788  None  
2008  None

In [ ]:
twitter_archive.nunique()

tweet_id                      2356
in_reply_to_status_id           77
in_reply_to_user_id             31
timestamp                     2356
source                           4
text                          2356
retweeted_status_id            181
retweeted_status_user_id        25
retweeted_status_timestamp     181
expanded_urls                 2218
rating_numerator                40
rating_denominator              18
name                           957
doggo                            2
floofer                          2
pupper                           2
puppo                            2
dtype: int64

In [ ]:
twitter_archive.rating_numerator.value_counts()
#the numerator and demoninator are silly and the entire purpose of the account. 
#Getting rid of "outliers" wouldn't help analytically.

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
2         9
1         9
75        2
15        2
420       2
0         2
80        1
144       1
17        1
26        1
20        1
121       1
143       1
44        1
60        1
45        1
50        1
99        1
204       1
1776      1
165       1
666       1
27        1
182       1
24        1
960       1
84        1
88        1
Name: rating_numerator, dtype: int64

In [ ]:
twitter_archive.rating_denominator.value_counts()
#Typically the denominator is 10, to be less than the numerator. It appears
#entires with higher denominators are almost always multiple dogs.

10     2333
11        3
50        3
20        2
80        2
70        1
7         1
15        1
150       1
170       1
0         1
90        1
40        1
130       1
110       1
16        1
120       1
2         1
Name: rating_denominator, dtype: int64

In [ ]:
#It looks like the vast majority of the denomiators are 10. 
#Checking to see if the others may have been written in error or just in jest. 
#Just the original tweets, no retweets.
rank_mask = (twitter_archive.rating_denominator != 10) & \
            twitter_archive.in_reply_to_status_id.isna() & \
            twitter_archive.retweeted_status_id.isna()

col_mask = ['text', 'tweet_id', 'rating_numerator', 'rating_denominator']
twitter_archive[rank_mask][col_mask]

text  \
433                                                                         The floofs have been released I repeat the floofs have been released. 84/70 https://t.co/NIYC820tmd   
516   Meet Sam. She smiles 24/7 &amp; secretly aspires to be a reindeer. \nKeep Sam smiling by clicking and sharing this link:\nhttps://t.co/98tB8y7y7t https://t.co/LouL5vdvxx   
902                                                                                              Why does this never happen at my front door... 165/150 https://t.co/HmwrdfEfUE   
1068                               After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https://t.co/XAVDNDaVgQ   
1120                                                  Say hello to this unbelievably well behaved squad of doggos. 204/170 would try to pet all at once https://t.co/yGQI3He3xv   
1165                                                                                                           Happy 4/20 from the squad! 13/10 for all https://t.co/eV1diwds8a   
1202                                                      This is Bluebert. He just saw that both #FinalFur match ups are split 50/50. Amazed af. 11/10 https://t.co/Kky1DPG4iq   
1228                                                                              Happy Saturday here's 9 puppers on a bench. 99/90 good work everybody https://t.co/mpvaVxKmc1   
1254                                                               Here's a brigade of puppers. All look very prepared for whatever happens next. 80/80 https://t.co/0eb7R1Om12   
1274                         From left to right:\nCletus, Jerome, Alejandro, Burp, &amp; Titson\nNone know where camera is. 45/50 would hug all at once https://t.co/sedre1ivTK   
1351                                                                                         Here is a whole flock of puppers.  60/50 I'll take the lot https://t.co/9dpcw6MdWa   
1433                                                                               Happy Wednesday here's a bucket of pups. 44/40 would pet all at once https://t.co/HppvrYuamZ   
1635                               Someone help the girl is being mugged. Several are distracting her while two steal her shoes. Clever puppers 121/110 https://t.co/1zfnTJLt55   
1662                                This is Darrel. He just robbed a 7/11 and is in a high speed police chase. Was just spotted by the helicopter 10/10 https://t.co/7EsP8LmSp5   
1779                                                                                                     IT'S PUPPERGEDDON. Total of 144/120 ...I think https://t.co/ZanVtAtvIq   
1843                                                                Here we have an entire platoon of puppers. Total score: 88/80 would pet all at once https://t.co/y93p6FLvVw   
2335                                   This is an Albanian 3 1/2 legged  Episcopalian. Loves well-polished hardwood flooring. Penis on the collar. 9/10 https://t.co/d9NcXFKwLv   

                tweet_id  rating_numerator  rating_denominator  
433   820690176645140481                84                  70  
516   810984652412424192                24                   7  
902   758467244762497024               165                 150  
1068  740373189193256964                 9                  11  
1120  731156023742988288               204                 170  
1165  722974582966214656                 4                  20  
1202  716439118184652801                50                  50  
1228  713900603437621249                99                  90  
1254  710658690886586372                80                  80  
1274  709198395643068416                45                  50  
1351  704054845121142784                60                  50  
1433  697463031882764288                44                  40  
1635  684222868335505415               121                 110  
1662  682962037429899265              

Most of these are groups of dogs, howver some are indeed incorrectly scrapped.

`810984652412424192` - Not a rating, just a link to click. Just drop.

`740373189193256964` - 9/11 Needs changed to 14/10

`722974582966214656` - Change 4/20 to 13/10

`716439118184652801` - Change 50/50 to 11/10

`682962037429899265` - Change 7/11 to 10/10

`666287406224695296` - Change 1/2 to 9/10.

In [ ]:
twitter_archive.duplicated().sum()

0

In [ ]:
twitter_archive.tweet_id.duplicated().sum() 
#There aren't any duplicated ID's for combination purposes later on.

0

In [ ]:
twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [ ]:
x = twitter_archive.doggo.value_counts()
y = twitter_archive.pupper.value_counts()
z = twitter_archive.puppo.value_counts()
a = twitter_archive.floofer.value_counts()
print(x, y, z, a)
#Not all dogs are "rated", the vast majority are not. 
#It also appears there are multiple versions of each style. Doggo, doggo, doggos, etc.

None     2259
doggo      97
Name: doggo, dtype: int64 None      2099
pupper     257
Name: pupper, dtype: int64 None     2326
puppo      30
Name: puppo, dtype: int64 None       2346
floofer      10
Name: floofer, dtype: int64


In [ ]:
twitter_archive.name.value_counts() 
#the top name "a" seems like a typo or placeholder. It's unlikely anyone
#would name thier dog that, let alone 55 seperate people.

None          745
a              55
Charlie        12
Cooper         11
Lucy           11
             ... 
Dex             1
Ace             1
Tayzie          1
Grizzie         1
Christoper      1
Name: name, Length: 957, dtype: int64

In [ ]:
# All the non capitalzied "names" and/or letters
non_cap = twitter_archive.name.str.contains('^[a-z]', regex = True)
twitter_archive[non_cap].name.value_counts().sort_index()

a               55
actually         2
all              1
an               7
by               1
getting          2
his              1
incredibly       1
infuriating      1
just             4
life             1
light            1
mad              2
my               1
not              2
officially       1
old              1
one              4
quite            4
space            1
such             1
the              8
this             1
unacceptable     1
very             5
Name: name, dtype: int64

In [ ]:
name_check = twitter_archive[non_cap].T
name_check.head()
#After viewing these they all truely don't contain a name, just a silly description.
#I'll leave them in, but use a query when doing an analysis.

22    \
tweet_id                                                                               887517139158093824   
in_reply_to_status_id                                                                                 NaN   
in_reply_to_user_id                                                                                   NaN   
timestamp                                                                       2017-07-19 03:39:09 +0000   
source                 <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                     56    \
tweet_id                                                                               881536004380872706   
in_reply_to_status_id                                                                                 NaN   
in_reply_to_user_id                                                                                   NaN   
timestamp                                                                       2017-07-02 15:32:16 +0000   
source                 <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                     118   \
tweet_id                                                                               869988702071779329   
in_reply_to_status_id                                                                                 NaN   
in_reply_to_user_id                                                                                   NaN   
timestamp                                                                       2017-05-31 18:47:24 +0000   
source                 <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                     169   \
tweet_id                                                                               859196978902773760   
in_reply_to_status_id                                                                                 NaN   
in_reply_to_user_id                                                                                   NaN   
timestamp                                                                       2017-05-02 00:04:57 +0000   
source                 <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                     193   \
tweet_id                                                                               855459453768019968   
in_reply_to_status_id                                                                                 NaN   
in_reply_to_user_id                                                                                   NaN   
timestamp                                                                       2017-04-21 16:33:22 +0000   
source                 <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                     335   \
tweet_id                                                               832645525019123713   
in_reply_to_status_id                                                                 NaN   
in_reply_to_user_id                                                                   NaN   
timestamp                                                       2017-02-17 17:38:57 +0000   
source                 <a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>   

                                                                                                     369   \
tweet_id                                                                               828650029636317184   
in_reply_to_status_id                            

## `image_pred` Assessment

In [ ]:
image_pred.sample(5)

tweet_id                                          jpg_url  \
1242  747204161125646336  https://pbs.twimg.com/media/Cl6aOBhWEAALuti.jpg   
1337  758405701903519748  https://pbs.twimg.com/media/CoZl9fXWgAMox0n.jpg   
110   667886921285246976  https://pbs.twimg.com/media/CUTPnPCW4AI7R0y.jpg   
2056  888554962724278272  https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
1034  711732680602345472  https://pbs.twimg.com/media/CeCVGEbUYAASeY4.jpg   

      img_num                        p1   p1_conf  p1_dog            p2  \
1242        2                      coil  0.533699   False        dugong   
1337        4  Chesapeake_Bay_retriever  0.702954    True        laptop   
110         1                Pomeranian  0.800432    True      Pekinese   
2056        3            Siberian_husky  0.700377    True    Eskimo_dog   
1034        3                     dingo  0.366875   False  Ibizan_hound   

       p2_conf  p2_dog           p3   p3_conf  p3_dog  
1242  0.087959   False  rain_barrel  0.039221   False  
1337  0.092277   False     notebook  0.032727   False  
110   0.168445    True    Chihuahua  0.008950    True  
2056  0.166511    True     malamute  0.111411    True  
1034  0.334929    True   Eskimo_dog  0.073876    True

In [ ]:
image_pred.nunique()
#There aren't the same number of URL's as tweet_id's. Suggesting not all the tweets had images, and are perhaps retweets or replies.

tweet_id    2075
jpg_url     2009
img_num        4
p1           378
p1_conf     2006
p1_dog         2
p2           405
p2_conf     2004
p2_dog         2
p3           408
p3_conf     2006
p3_dog         2
dtype: int64

In [ ]:
image_pred['jpg_url'].duplicated().sum()
#There are 66 image urls duplicated, so possibly multiple predictions of the same image.

66

In [ ]:
image_pred['jpg_url'].duplicated().value_counts()

False    2009
True       66
Name: jpg_url, dtype: int64

In [ ]:
image_pred.duplicated().sum()

0

In [ ]:
image_pred.tweet_id.duplicated().sum()
#no duplicate ID's for combining purposes later on.

0

In [ ]:
image_pred.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [ ]:
image_pred.p1.value_counts()

golden_retriever      150
Labrador_retriever    100
Pembroke               89
Chihuahua              83
pug                    57
                     ... 
pillow                  1
carousel                1
bald_eagle              1
lorikeet                1
orange                  1
Name: p1, Length: 378, dtype: int64

In [ ]:
image_pred.p2.value_counts()

Labrador_retriever    104
golden_retriever       92
Cardigan               73
Chihuahua              44
Pomeranian             42
                     ... 
medicine_chest          1
quail                   1
horse_cart              1
waffle_iron             1
bagel                   1
Name: p2, Length: 405, dtype: int64

In [ ]:
image_pred.p3.value_counts()

Labrador_retriever    79
Chihuahua             58
golden_retriever      48
Eskimo_dog            38
kelpie                35
                      ..
ox                     1
assault_rifle          1
axolotl                1
pot                    1
banana                 1
Name: p3, Length: 408, dtype: int64

## `tweet_json` Assessment

In [ ]:
tweet_json.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   tweet_id        2354 non-null   int64
 1   retweet_count   2354 non-null   int64
 2   favorite_count  2354 non-null   int64
dtypes: int64(3)
memory usage: 55.3 KB


### Quality issues
1. The Datatype for dog styles should be category. (twitter_archive)

2. Timestamp should be in datetime format, not object in twitter_archive. Also remove the “+0000”. (twitter_archive)

3. Some of the tweets have inccorect values for ranking, as other fractions were listed before the correct ranking was in the tweet. 

4. There are versions of dog stlyes that are technically the same, but written differently. Doggo, doggo, doggos, etc. (imgage_pred)

5. There are a few jpg_url duplicates. They appear to be related to retweets. (Image Pred)

6. Not every dog gets a doggo “rank”. Can’t use it for an overall analysis, might be able to do a mini analysis on just the ranks.

7. There are retweets and replies, I only want the original ratings to not mess up the analysis down the line. (twitter_archive)

8. Some of the names look wrong for some of the dogs. “A” is not a name. (twitter_archive)


### Tidiness issues
1. The URL's in `twitter_archive` are still in HTML format. It would be easier to read and visually assess what's happening if I extract only the text and remove all the uneeded HTML.

2.Dog “rank” (doggo, puppo, floof, etc.) should just be one column.

## Cleaning Data

Cleaning will be done in a Define, Code, Test framework. Where I define the problem, code the solution, then test it to make sure it works.

In [ ]:
# Make copies of original pieces of data
twitter_archive_c = twitter_archive.copy()
image_pred_c = image_pred.copy()
tweet_json_c = tweet_json.copy()

# Cleaning `twitter_archive_c`

### Issue #1: In `twitter_archive_c` "timestamp" should be in datetime format.

#### Define
Change timestamp from object to datetime64.

#### Code

In [ ]:
twitter_archive_c.timestamp = pd.to_datetime(twitter_archive_c.timestamp)

#### Test

In [ ]:
twitter_archive_c.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype              
---  ------                      --------------  -----              
 0   tweet_id                    2356 non-null   int64              
 1   in_reply_to_status_id       78 non-null     float64            
 2   in_reply_to_user_id         78 non-null     float64            
 3   timestamp                   2356 non-null   datetime64[ns, UTC]
 4   source                      2356 non-null   object             
 5   text                        2356 non-null   object             
 6   retweeted_status_id         181 non-null    float64            
 7   retweeted_status_user_id    181 non-null    float64            
 8   retweeted_status_timestamp  181 non-null    object             
 9   expanded_urls               2297 non-null   object             
 10  rating_numerator            2356 non-null   int64           

## Issue # 2: We only want original content, not retweets.

#### Define: From `twitter_archive_c` we need to remove replies and retweets. 

#### Code

In [ ]:
#Remove any rows not containing any NaN values. Then, drop the uneeded columns.
replies = (~twitter_archive_c.in_reply_to_status_id.isnull())
#There are 76 total.

twitter_archive_c = twitter_archive_c[~replies]

In [ ]:
rt = (~twitter_archive_c.retweeted_status_id.isnull())
# There are 181 total.

twitter_archive_c = twitter_archive_c[~rt]

In [ ]:
twitter_archive_c.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype              
---  ------                      --------------  -----              
 0   tweet_id                    2097 non-null   int64              
 1   in_reply_to_status_id       0 non-null      float64            
 2   in_reply_to_user_id         0 non-null      float64            
 3   timestamp                   2097 non-null   datetime64[ns, UTC]
 4   source                      2097 non-null   object             
 5   text                        2097 non-null   object             
 6   retweeted_status_id         0 non-null      float64            
 7   retweeted_status_user_id    0 non-null      float64            
 8   retweeted_status_timestamp  0 non-null      object             
 9   expanded_urls               2094 non-null   object             
 10  rating_numerator            2097 non-null   int64           

In [ ]:
#I know I can make this all one line for easier coding, but for whatever reason,
#I keep getting KeyErrors for any other way I code it.
twitter_archive_c = twitter_archive_c.drop(['retweeted_status_id'], axis=1)
twitter_archive_c = twitter_archive_c.drop(['retweeted_status_user_id'], axis=1)
twitter_archive_c = twitter_archive_c.drop(['retweeted_status_timestamp'], axis=1)
twitter_archive_c = twitter_archive_c.drop(['in_reply_to_status_id'], axis=1)
twitter_archive_c = twitter_archive_c.drop(['in_reply_to_user_id'], axis=1)

#### Test

In [ ]:
twitter_archive_c.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   tweet_id            2097 non-null   int64              
 1   timestamp           2097 non-null   datetime64[ns, UTC]
 2   source              2097 non-null   object             
 3   text                2097 non-null   object             
 4   expanded_urls       2094 non-null   object             
 5   rating_numerator    2097 non-null   int64              
 6   rating_denominator  2097 non-null   int64              
 7   name                2097 non-null   object             
 8   doggo               2097 non-null   object             
 9   floofer             2097 non-null   object             
 10  pupper              2097 non-null   object             
 11  puppo               2097 non-null   object             
dtypes: datetime64[ns, UTC](1), int64(3

#### Test 

### Issue #3: In `twitter_archive_c` there are dog names that aren't actual names.

- Luckily, it appears all the actual names are all capitalized. It appears we can run a check to see if any uncapitalized letters or words are actually just accidental additions to the name column. I'll do a manual check to see if there are any actual names, if not I'll just make them NaN.

#### Code

In [ ]:
# All the non capitalzied "names" and/or letters
non_cap = twitter_archive_c.name.str.contains('^[a-z]', regex = True)
twitter_archive_c[non_cap].name.value_counts().sort_index()

a               55
actually         2
all              1
an               6
by               1
getting          2
his              1
incredibly       1
infuriating      1
just             3
life             1
light            1
mad              1
my               1
not              2
officially       1
old              1
one              4
quite            3
space            1
such             1
the              8
this             1
unacceptable     1
very             4
Name: name, dtype: int64

In [ ]:
twitter_archive_c.loc[non_cap, 'name'] = "None"
twitter_archive_c['name'].replace('None', np.nan, inplace=True)

#### Test

In [ ]:
twitter_archive_c.name.head(10)

0     Phineas
1       Tilly
2      Archie
3       Darla
4    Franklin
5         NaN
6         Jax
7         NaN
8        Zoey
9      Cassie
Name: name, dtype: object

In [ ]:
twitter_archive_c.name.value_counts()

Charlie       11
Lucy          11
Oliver        10
Cooper        10
Penny          9
              ..
Shelby         1
Sephie         1
Bonaparte      1
Wishes         1
Christoper     1
Name: name, Length: 929, dtype: int64

### Issue # 4 Not every dog gets a doggo “type”. And there are redundant columns we can remove for tidyness.

#### Define

- take the three dog stages and make a new column with the combined data named `dog_stage`. Then drop the 3 original columns, leaving only the cleaned one.

- There are 11 tweets in the dataset that have more than one dog stage. In order to avoid dropping them entirely it's going to be the first rank in ascending order. The order being doggo, floofer, pupper, puppo.

#### Code

In [ ]:
#Trying to catch any version of the word, capitalized or not
twitter_archive_c['doggo'] = twitter_archive_c['text'].str.extract('([Dd]+[Oo]+[Gg]+[Oo]+[Ss]*)')
twitter_archive_c['floofer'] = twitter_archive_c['text'].str.extract('([Ff]+[Ll]+[Oo]+[Ff]+[Ee]+[Rr]+[Ss]*)')
twitter_archive_c['pupper'] = twitter_archive_c['text'].str.extract('([Pp]+[Uu]+[Pp]+[Ee]+[Rr]+[Ss]*)')
twitter_archive_c['puppo'] = twitter_archive_c['text'].str.extract('([Pp]+[Uu]+[Pp]+[Oo]+[Ss]*)')

#### Test

In [ ]:
twitter_archive_c['doggo'].value_counts()

doggo     75
doggos     8
Doggo      7
DOGGO      1
Name: doggo, dtype: int64

In [ ]:
twitter_archive_c['puppo'].value_counts()

puppo     28
Puppo      1
puppos     1
Name: puppo, dtype: int64

In [ ]:
twitter_archive_c['floofer'].value_counts()

Floofer    6
floofer    4
Name: floofer, dtype: int64

In [ ]:
twitter_archive_c['pupper'].value_counts()

pupper     221
puppers     22
Pupper       8
PUPPER       4
Name: pupper, dtype: int64

In [ ]:
#Combine the dog type colmns into one column.
twitter_archive_c['dog_stage'] = twitter_archive_c.doggo.fillna(twitter_archive_c.floofer.fillna(twitter_archive_c.pupper.fillna(twitter_archive_c.puppo)))

In [ ]:
twitter_archive_c[['dog_stage', 'doggo', 'floofer', 'pupper', 'puppo']].sample(20)

dog_stage doggo floofer   pupper  puppo
1432       NaN   NaN     NaN      NaN    NaN
1578       NaN   NaN     NaN      NaN    NaN
969        NaN   NaN     NaN      NaN    NaN
1145       NaN   NaN     NaN      NaN    NaN
920        NaN   NaN     NaN      NaN    NaN
1210       NaN   NaN     NaN      NaN    NaN
703        NaN   NaN     NaN      NaN    NaN
1440       NaN   NaN     NaN      NaN    NaN
208        NaN   NaN     NaN      NaN    NaN
1765       NaN   NaN     NaN      NaN    NaN
1826   puppers   NaN     NaN  puppers    NaN
901        NaN   NaN     NaN      NaN    NaN
713      puppo   NaN     NaN      NaN  puppo
1652    pupper   NaN     NaN   pupper    NaN
820        NaN   NaN     NaN      NaN    NaN
1493    pupper   NaN     NaN   pupper    NaN
315        NaN   NaN     NaN      NaN    NaN
133        NaN   NaN     NaN      NaN    NaN
1406       NaN   NaN     NaN      NaN    NaN
2099       NaN   NaN     NaN      NaN    NaN

In [ ]:
twitter_archive_c.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   tweet_id            2097 non-null   int64              
 1   timestamp           2097 non-null   datetime64[ns, UTC]
 2   source              2097 non-null   object             
 3   text                2097 non-null   object             
 4   expanded_urls       2094 non-null   object             
 5   rating_numerator    2097 non-null   int64              
 6   rating_denominator  2097 non-null   int64              
 7   name                1390 non-null   object             
 8   doggo               91 non-null     object             
 9   floofer             10 non-null     object             
 10  pupper              255 non-null    object             
 11  puppo               30 non-null     object             
 12  dog_stage           374 non-null  

In [ ]:
# Remove the now uneeded columns.
twitter_archive_c.drop(['doggo', 
                    'floofer', 
                    'pupper', 
                    'puppo'], axis=1, inplace=True)

In [ ]:
# Making sure the dog_style datatype is categorical.
twitter_archive_c['dog_stage'] = twitter_archive_c['dog_stage'].astype('category')

#### Test

In [ ]:
twitter_archive_c.sample(5)

tweet_id                 timestamp  \
481   815639385530101762 2017-01-01 19:22:38+00:00   
837   767754930266464257 2016-08-22 16:06:54+00:00   
609   797545162159308800 2016-11-12 21:02:38+00:00   
789   773985732834758656 2016-09-08 20:45:53+00:00   
1482  693267061318012928 2016-01-30 02:58:42+00:00   

                                                                                  source  \
481   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
837   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
609   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
789   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1482                     <a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>   

                                                                                                                                              text  \
481                   This is Titan. His nose is quite chilly. Requests to return to the indoors. 12/10 would boop to warm https://t.co/bLZuOh9sKy   
837           This is Philbert. His toilet broke and he doesn't know what to do. Trying not to panic. 11/10 furustrated af https://t.co/Nb68IsVb9O   
609          This is Cassie. She steals things. Guilt increases slightly each time. 12/10 would forgive almost immediately https://t.co/Ia19irLwyB   
789   Meet Winnie. She just made awkward eye contact with the driver beside her. Poor pupper panicked. 11/10 would comfort https://t.co/RFWtDqTnAz   
1482                            This is Oscar. He can wave. Friendly af. 12/10 would totally wave back (IG: Oscar.is.bear) https://t.co/waN6EW0wfM   

                                                                                                                                                                                                                                                        expanded_urls  \
481                                                                                                                                                                                                   https://twitter.com/dog_rates/status/815639385530101762/photo/1   
837                                                                                                                                                                                                   https://twitter.com/dog_rates/status/767754930266464257/photo/1   
609   https://twitter.com/dog_rates/status/797545162159308800/photo/1,https://twitter.com/dog_rates/status/797545162159308800/photo/1,https://twitter.com/dog_rates/status/797545162159308800/photo/1,https://twitter.com/dog_rates/status/797545162159308800/photo/1   
789   https://twitter.com/dog_rates/status/773985732834758656/photo/1,https://twitter.com/dog_rates/status/773985732834758656/photo/1,https://twitter.com/dog_rates/status/773985732834758656/photo/1,https://twitter.com/dog_rates/status/773985732834758656/photo/1   
1482                                                                                                                                                                                                                                    https://vine.co/v/i5n2irFUYWv   

      rating_numerator  rating_denominator      name dog_stage  
481                 12                  10     Titan       NaN  
837                 11                  10  Philbert       NaN  
609                 12                  10    Cassie       NaN  
789                 11                  10    Winnie    pupper  
1482                12                  10     Oscar       NaN

In [ ]:
twitter_archive_c.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   tweet_id            2097 non-null   int64              
 1   timestamp           2097 non-null   datetime64[ns, UTC]
 2   source              2097 non-null   object             
 3   text                2097 non-null   object             
 4   expanded_urls       2094 non-null   object             
 5   rating_numerator    2097 non-null   int64              
 6   rating_denominator  2097 non-null   int64              
 7   name                1390 non-null   object             
 8   dog_stage           374 non-null    category           
dtypes: category(1), datetime64[ns, UTC](1), int64(3), object(4)
memory usage: 150.1+ KB


## Issue # 5: The Source column has the HTML tags in it. If we remove them it will make the table much easier to read.

In [ ]:
#The current state of things
twitter_archive_c.source.value_counts()

<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     1964
<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                          91
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                       31
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>      11
Name: source, dtype: int64

#### Code

In [ ]:
twitter_archive_c['source'] = twitter_archive_c['source'].str.extract('^<a.+>(.+)</a>$')

In [ ]:
#While we're at it, going to remove the `expanded_urls` column for tidyness.
twitter_archive_c.drop(['expanded_urls'], axis=1, inplace=True)

### Test

In [ ]:
twitter_archive_c.source.value_counts() #Much better

Twitter for iPhone     1964
Vine - Make a Scene      91
Twitter Web Client       31
TweetDeck                11
Name: source, dtype: int64

### Issue # 6 : Some of the numerators and denomiators are incorrect because the scraped data shows fractions used in the tweet instead of the correct rating fraction at the end of the tweet.


* `810984652412424192` - Not a rating, just a link to click. Just drop.
* `740373189193256964` - 9/11 Needs changed to 14/10
* `722974582966214656` - Change 4/20 to 13/10
* `716439118184652801` - Change 50/50 to 11/10
* `682962037429899265` - Change 7/11 to 10/10
* `666287406224695296` - Change 1/2 to 9/10.

#### Code

In [ ]:
fix = twitter_archive_c.tweet_id == 740373189193256964
twitter_archive_c.loc[fix, 'rating_numerator'] = 14
twitter_archive_c.loc[fix, 'rating_denominator'] = 10

In [ ]:
fix = twitter_archive_c.tweet_id == 722974582966214656
twitter_archive_c.loc[fix, 'rating_numerator'] = 13
twitter_archive_c.loc[fix, 'rating_denominator'] = 10

In [ ]:
fix = twitter_archive_c.tweet_id == 716439118184652801
twitter_archive_c.loc[fix, 'rating_numerator'] = 11
twitter_archive_c.loc[fix, 'rating_denominator'] = 10

In [ ]:
fix = twitter_archive_c.tweet_id == 682962037429899265
twitter_archive_c.loc[fix, 'rating_numerator'] = 10
twitter_archive_c.loc[fix, 'rating_denominator'] = 10

In [ ]:
fix = twitter_archive_c.tweet_id == 666287406224695296
twitter_archive_c.loc[fix, 'rating_numerator'] = 9
twitter_archive_c.loc[fix, 'rating_denominator'] = 10

In [ ]:
#Dropping the tweet that is just a link.
twitter_archive_c.drop(index=516, inplace=True)

#### Test

In [ ]:
check = ['tweet_id', 'rating_numerator', 'rating_denominator']
twitter_archive_c[(twitter_archive_c.tweet_id == 740373189193256964) | \
              (twitter_archive_c.tweet_id == 722974582966214656) | \
              (twitter_archive_c.tweet_id == 716439118184652801) | \
              (twitter_archive_c.tweet_id == 682962037429899265) | \
              (twitter_archive_c.tweet_id == 666287406224695296) ][check]

tweet_id  rating_numerator  rating_denominator
1068  740373189193256964                14                  10
1165  722974582966214656                13                  10
1202  716439118184652801                11                  10
1662  682962037429899265                10                  10
2335  666287406224695296                 9                  10

In [ ]:
twitter_archive_c.info()
#There is one less entry than before, the code for dropping the row worked

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2096 entries, 0 to 2355
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   tweet_id            2096 non-null   int64              
 1   timestamp           2096 non-null   datetime64[ns, UTC]
 2   source              2096 non-null   object             
 3   text                2096 non-null   object             
 4   rating_numerator    2096 non-null   int64              
 5   rating_denominator  2096 non-null   int64              
 6   name                1389 non-null   object             
 7   dog_stage           374 non-null    category           
dtypes: category(1), datetime64[ns, UTC](1), int64(3), object(3)
memory usage: 133.7+ KB


## Cleaning `image_pred`

### Issue # 7: There are pictures in the `image_pred` dataframe that are not of dogs.

#### Define: I'm going to query `image_pred` for any rows that do not contain at least one **True** value between the three dog guesses. This will leave only rows with correct dog guesses.

#### Code

In [ ]:
image_pred_c.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [ ]:
image_pred_c = image_pred_c.query('p1_dog == True | p2_dog == True | p3_dog == True')
image_pred_c.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1751 entries, 0 to 2073
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  1751 non-null   int64  
 1   jpg_url   1751 non-null   object 
 2   img_num   1751 non-null   int64  
 3   p1        1751 non-null   object 
 4   p1_conf   1751 non-null   float64
 5   p1_dog    1751 non-null   bool   
 6   p2        1751 non-null   object 
 7   p2_conf   1751 non-null   float64
 8   p2_dog    1751 non-null   bool   
 9   p3        1751 non-null   object 
 10  p3_conf   1751 non-null   float64
 11  p3_dog    1751 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 141.9+ KB


In [ ]:
#Double-checking to see if there are rows that do not contain any dog guesses at all.
image_pred_c.query('p1_dog == False & p2_dog == False & p3_dog == False')

Empty DataFrame
Columns: [tweet_id, jpg_url, img_num, p1, p1_conf, p1_dog, p2, p2_conf, p2_dog, p3, p3_conf, p3_dog]
Index: []

In [ ]:
# I'm going to set up a breed and confidence column using out p1, p1_conf etc.
#First I'll set up a sort of mask for breed predictions, then the order for confidence.
#Then create the columns from the first successful condition of them. 
needs = [(image_pred_c['p1_dog'] == True),
              (image_pred_c['p2_dog'] == True),
              (image_pred_c['p3_dog'] == True)]

# set the choice order based on the selection conditions for predicted breed
breed = [image_pred_c['p1'], 
                 image_pred_c['p2'],
                 image_pred_c['p3']]

# set the choice order for confidence level based on the selection conditions
conf = [image_pred_c['p1_conf'], 
                      image_pred_c['p2_conf'], 
                      image_pred_c['p3_conf']]

# select the predicted breed based on the first successful condition
image_pred_c['breeds'] = np.select(needs, breed, 
                                       default = 'none')

# select the predicted confidence level based on the first successful condition
image_pred_c['confidence'] = np.select(needs, conf, 
                                            default = 0)

### Test

In [ ]:
image_pred_c.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  \
0  0.156665    True    Shetland_sheepdog  0.061428    True   
1  0.074192    True  Rhodesian_ridgeback  0.072010    True   
2  0.138584    True           bloodhound  0.116197    True   
3  0.360687    True   miniature_pinscher  0.222752    True   
4  0.243682    True             Doberman  0.154629    True   

                   breeds  confidence  
0  Welsh_springer_spaniel    0.465074  
1                 redbone    0.506826  
2         German_shepherd    0.596461  
3     Rhodesian_ridgeback    0.408143  
4      miniature_pinscher    0.560311

In [ ]:
#Make the confidence a percent by multiplying it by 100.
image_pred_c.confidence = (image_pred_c.confidence * 100).astype(int)

In [ ]:

#Also going to remove the `img_num`, as it's not needed.
image_pred_c.drop(['img_num'], axis=1, inplace=True)

In [ ]:
image_pred_c.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

                       p1   p1_conf  p1_dog                  p2   p2_conf  \
0  Welsh_springer_spaniel  0.465074    True              collie  0.156665   
1                 redbone  0.506826    True  miniature_pinscher  0.074192   
2         German_shepherd  0.596461    True            malinois  0.138584   
3     Rhodesian_ridgeback  0.408143    True             redbone  0.360687   
4      miniature_pinscher  0.560311    True          Rottweiler  0.243682   

   p2_dog                   p3   p3_conf  p3_dog                  breeds  \
0    True    Shetland_sheepdog  0.061428    True  Welsh_springer_spaniel   
1    True  Rhodesian_ridgeback  0.072010    True                 redbone   
2    True           bloodhound  0.116197    True         German_shepherd   
3    True   miniature_pinscher  0.222752    True     Rhodesian_ridgeback   
4    True             Doberman  0.154629    True      miniature_pinscher   

   confidence  
0          46  
1          50  
2          59  
3          40  
4          56

## Issue # 8 There are duplicate `jpg_url` entries.

#### Define: 
I'm just going to drop any duplicates, as when visially assessed they appeared to all just be retweets.

#### Code

In [ ]:
image_pred_c.jpg_url.duplicated().sum()
# There are 66
image_pred_c.drop_duplicates(inplace=True)

#### Test

In [ ]:
image_pred_c.duplicated().sum()

0

In [ ]:
#Might as wekk drop the jpg_url column for tidyness.
image_pred_c.drop(['jpg_url'], axis=1, inplace=True)

## Cleaning `tweet_json_c`

-The only real thing is to make sure that the tweets not shared by both dataframes get dropped. This would make it easier to join it with the `twitter_archive_c` dataframe via `tweet_id`

In [ ]:
tweet_json.head()

tweet_id  retweet_count  favorite_count
0  892420643555336193           8853           39467
1  892177421306343426           6514           33819
2  891815181378084864           4328           25461
3  891689557279858688           8964           42908
4  891327558926688256           9774           41048

#### Code

In [ ]:
no_share = (~twitter_archive_c.tweet_id.isin(list(tweet_json_c.tweet_id)))
no_share.sum()
#Oh good, there aren't any.

0

In [ ]:
#Merging tweet_json with twittter_archive_c by tweet_id.
twitter_archive_c = pd.merge(twitter_archive_c, tweet_json,
                         on = 'tweet_id', how = 'inner')

In [ ]:
twitter_archive_c.head()

tweet_id                 timestamp              source  \
0  892420643555336193 2017-08-01 16:23:56+00:00  Twitter for iPhone   
1  892177421306343426 2017-08-01 00:17:27+00:00  Twitter for iPhone   
2  891815181378084864 2017-07-31 00:18:03+00:00  Twitter for iPhone   
3  891689557279858688 2017-07-30 15:58:51+00:00  Twitter for iPhone   
4  891327558926688256 2017-07-29 16:00:24+00:00  Twitter for iPhone   

                                                                                                                                                                 text  \
0                                                       This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU   
1  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
2                   This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB   
3                                                             This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ   
4  This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek https://t.co/AtUZn91f7f   

   rating_numerator  rating_denominator      name dog_stage  retweet_count  \
0                13                  10   Phineas       NaN           8853   
1                13                  10     Tilly       NaN           6514   
2                12                  10    Archie       NaN           4328   
3                13                  10     Darla       NaN           8964   
4                12                  10  Franklin       NaN           9774   

   favorite_count  
0           39467  
1           33819  
2           25461  
3           42908  
4           41048

In [ ]:
#The MegaMerge.
# merge the breed and confidence columns to archive
merge = ['tweet_id', 'breeds', 'confidence']
twitter_archive_c = pd.merge(twitter_archive_c, image_pred_c[merge],
                         on = 'tweet_id', how = 'inner')

#### Test

In [ ]:
twitter_archive_c.head()

tweet_id                 timestamp              source  \
0  892177421306343426 2017-08-01 00:17:27+00:00  Twitter for iPhone   
1  891815181378084864 2017-07-31 00:18:03+00:00  Twitter for iPhone   
2  891689557279858688 2017-07-30 15:58:51+00:00  Twitter for iPhone   
3  891327558926688256 2017-07-29 16:00:24+00:00  Twitter for iPhone   
4  891087950875897856 2017-07-29 00:08:17+00:00  Twitter for iPhone   

                                                                                                                                                                 text  \
0  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
1                   This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB   
2                                                             This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ   
3  This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek https://t.co/AtUZn91f7f   
4  Here we have a majestic great white breaching off South Africa's coast. Absolutely h*ckin breathtaking. 13/10 (IG: tucker_marlo) #BarkWeek https://t.co/kQ04fDDRmh   

   rating_numerator  rating_denominator      name dog_stage  retweet_count  \
0                13                  10     Tilly       NaN           6514   
1                12                  10    Archie       NaN           4328   
2                13                  10     Darla       NaN           8964   
3                12                  10  Franklin       NaN           9774   
4                13                  10       NaN       NaN           3261   

   favorite_count                    breeds  confidence  
0           33819                 Chihuahua          32  
1           25461                 Chihuahua          71  
2           42908        Labrador_retriever          16  
3           41048                    basset          55  
4           20562  Chesapeake_Bay_retriever          42

In [ ]:
twitter_archive_c.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1665 entries, 0 to 1664
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   tweet_id            1665 non-null   int64              
 1   timestamp           1665 non-null   datetime64[ns, UTC]
 2   source              1665 non-null   object             
 3   text                1665 non-null   object             
 4   rating_numerator    1665 non-null   int64              
 5   rating_denominator  1665 non-null   int64              
 6   name                1185 non-null   object             
 7   dog_stage           284 non-null    category           
 8   retweet_count       1665 non-null   int64              
 9   favorite_count      1665 non-null   int64              
 10  breeds              1665 non-null   object             
 11  confidence          1665 non-null   int32              
dtypes: category(1), datetime64[ns, UTC

In [ ]:
twitter_archive_c.to_csv('twitter_archive_master.csv', index=False)

In [ ]:
image_pred_c.to_csv('image_pred_master.csv', index=False)

### Resources:
https://stackoverflow.com/questions/22551403/python-pandas-filtering-out-nan-from-a-data-selection-of-a-column-of-strings

https://stackoverflow.com/questions/63148009/remove-0000-utc-offset-from-timestamp-in-python-pandas

https://stackoverflow.com/questions/23743460/replace-none-with-nan-in-pandas-dataframe

https://stackoverflow.com/questions/43387467/how-to-remove-seconds-from-datetime 

https://stackoverflow.com/questions/43772362/how-to-print-a-specific-row-of-a-pandas-dataframe

https://pandas.pydata.org/docs/reference/api/pandas.melt.html
